onnxruntime/core/providers/nnapi/nnapi_builtin/builders/op_builder.cc

```diff
@@ -1859,8 +1878,16 @@ Status ConcatOpBuilder::AddToModelBuilderImpl(ModelBuilder& model_builder, const
   const auto& input0 = inputs[0].node_arg.Name();
   const auto node_input_size = inputs.size();
 
-  // First if the inputs are uint8, we need verify all the inputs have same scale and zero points
-  if (operand_types.at(input0).type == android::nn::wrapper::Type::TENSOR_QUANT8_ASYMM) {
+  bool is_quant_op = IsQuantizedOp(node_unit);
+
+  // First if the inputs are uint8 and this is not a quantized Concat,
+  // we need verify all the inputs have same scale and zero points
+  // ONNX allows Concat (not QlinearConcat, not QDQ concat) run directly on uint8 without scales and zps,
+  // However, for NNAPI, all int8 inputs need scale and zero points cannot be 0
+  // so we need to use the scales and zps from the NNAPI input directly, there is no easy way to get the input
```

Comment:
The comment suggests there should be a check that the value is not zero, but the code doesn't do that. Should it? I don't see this requirement in the NNAPI doco [here](https://developer.android.com/ndk/reference/group/neural-networks#group___neural_networks_1ggaabbe492c60331b13038e39d4207940e0a44cbea825c4b224dd3ea757e9b1f65ed)

Reply:
The scale is required to be >0, 
https://android.googlesource.com/platform/frameworks/ml/+/master/nn/common/Validation.cpp#486

The 0 is the default scale value for NNAPI operand, however, with a valid model, this usually cannot be 0, so I did not check its value before, BTW, zero point can be 0 so the comment probably need little update

Reply:
updated the comment.

Reply:
Thanks for the info. I assume in some places '0' was used as unset (otherwise all the values will be == zp). Interestingly TF seems to convert a scale of 0 to 1. 

https://github.com/tensorflow/tensorflow/blob/7737c518a864e54be9b676fe063436ccbbef21b9/tensorflow/lite/delegates/nnapi/nnapi_delegate.cc#L468-L471